feat: add 'mousescroll' option

JaySandhu opened the pull request:
Building on https://github.com/neovim/neovim/pull/1288 by @cepko33  and on https://github.com/neovim/neovim/pull/10954 by @qstrahl.
Close #11738

Adds a mousescroll option which I understand now has the green light.

**Motivation**
Scrolling with a trackpad feels off with the current hardcoded value. Remapping the scroll wheel keys to \<C-Y> and \<C-E>, as is commonly suggested, is not a great alternative, you can no longer scroll inactive windows.

The new option defaults to 1, which is different from the old hardcoded value of 3, but I think is a more reasonable default. This can be changed if needed.

Horizontal scrolling still uses a hardcoded value (6 I believe).

- [x] Add the option
- [x] Document the option
- [x] Test the option manually
- [x] Create a new automated test
- [x] Update `:help scroll-mouse-wheel`
- [x] Update `:help vim-differences`

JaySandhu committed:
feat: add 'mousescroll' option

Add 'mousescroll' option to control how many lines to scroll by when a
mouse wheel keycode is received. The mousescroll option controls both
horizontal and vertical scrolling. The option is a string in the format:

    set mousescroll=direction:count,direction:count

Where direction is either 'ver' or 'hor', and count is a non negative
integer. If a direction is omitted, a default value is used. The default
values remain unchanged, that is 3 for vertical scrolling, and 6 for
horizontal scrolling. As such, the mousescroll default is "ver:3,hor:6".

Add mousescroll documentation
 - Add option documentation in options.txt
 - Add brief summary in quickref.txt

Update :help scroll-mouse-wheel
 - Mention mousescroll option as a means of controlling scrolling.
 - Remove obsolete suggestion to map scroll wheel keys to <C-U> to
   scroll by a single line -- users should prefer the mousescroll option.
 - Add some information about the consequences of remapping scroll wheel
   keys (they lose their magic ability to affect inactive windows).

Update :help vim-differences
 - Add brief mousescroll summary under Options

Add mousescroll tests
 - Test option validation
 - Test default mousescroll value and behavior
 - Test fallback to default values
 - Test mouse vertical and horizontal scrolling in normal mode
 - Test mouse vertical and horizontal scrolling in insert mode

justinmk commented:
Thanks for picking this up. Needs some tests; I added the checklist from https://github.com/neovim/neovim/pull/10954 to your description.

> Horizontal scrolling still uses a hardcoded value (6 I believe).

Will we need something like...

    :set mousescroll=vert:42,hor:42

JaySandhu commented:
> Will we need something like...
>
> ```
> :set mousescroll=ver:42,hor:42
> ```

I like that idea. It would make for a more complete solution. I'll rework the PR tomorrow.

Do we want to force users to set both directions? Or should we allow them to omit either `ver` or `hor` (in which case we can fallback to some default value). To put my two cents in, I think that we should support omitted directions.

JaySandhu commented:
I gave it a shot. The new approach is more complete, but also, more complex. The `mousescroll` option is now a string in the following format:

```
set mousescroll=direction:count,direction:count
```

Where direction is either 'ver' or 'hor', and count is a non negative integer.

Some points that may be contentious:

* Setting a direction twice results in an error. For example, `set mousescroll="ver:1,ver:2"` is rejected.
* A scroll value of 0 is accepted. Setting a direction to 0 disables scrolling along that axis.
* If a direction is omitted, the default value is used. For example:

`set mousescroll=ver:1,hor:3` - Sets vertical scrolling to 1, horizontal scrolling to 3.
`set mousescroll=ver:1` - Sets vertical scrolling to 1, horizontal scrolling to the default value (6).
`set mousescroll=` - Error, both directions were omitted.

Do we want to go with the string option or the number option from earlier? I'm okay with either, having the ability to control vertical scrolling gets us 95% of the way there. Once it's settled, I'll update the docs and add tests.

justinmk commented:
All of those choices sound fine. 

> Do we want to go with the string option or the number option from earlier?

not sure what you mean. The `set mousescroll=direction:count,direction:count` form looks great.

justinmk commented on src/nvim/option.c:
```diff
@@ -2557,6 +2557,76 @@ static bool valid_spellfile(const char_u *val)
+  if (horizontal == -1) {
+    p_mousescroll_hor = MOUSESCROLL_HOR_DFLT;
+  } else {
+    p_mousescroll_hor = horizontal;
```
nit: prefer ternaries for these simple if-else cases, it results in much shorter code without a loss of clarity.

JaySandhu replied:
Valid point, I made the change 👍

justinmk reviewed (approved):
looks great so far, looking forward to this! 

p.s. For faster feedback our chat channel  https://gitter.im/neovim/neovim can be useful.

JaySandhu commented:
Wrote a test and updated the docs. I think that covers it. The documentation might need some copy editing though. The test works by scrolling and observing changes in the cursor position, figured we could get away with not using `ui.screen`.

jamessan commented on src/nvim/option.c:
```diff
@@ -2568,6 +2568,69 @@ static bool valid_spellfile(const char_u *val)
+    }
+
+    errno = 0;
+    *direction = strtol((char *)string + 4, NULL, 10);
```
There's a family of `getdigits` functions in src/nvim/charset.c that encapsulate this sort of functionality and error-checking.  Probably better to use the appropriate one of those.

JaySandhu replied:
I had a look, I think the best way to do it is something like this:
```
string += 4;
*direction = getdigits_int(&string, false, -1);

// Num options are generally kept within the signed int range.
// We know this number won't be negative because we've already checked for
// a minus sign. We'll allow 0 as a means of disabling mouse scrolling.
if (*direction == -1) {
  return e_invarg;
}
```
Idk if that's helping readability, but it's your call.

overdrivenpotato commented:
Is there something blocking this PR from being merged?

dundargoc commented:
> Is there something blocking this PR from being merged?

Lack of manpower I'd guess :). @clason: who's most suited to tackle this you think?

clason commented:
What's blocking is the need for

1. addressing the review comment
2. rebasing on latest `master`

JaySandhu commented:
Happy to make `getdigits` change and rebase. 

For what it's worth, I don't think using `getdigits` is much of an improvement. Will do what it takes to get this merged though.

justinmk commented:
@JaySandhu gentle ping

JaySandhu commented:
Thanks for the ping, this had totally slipped my mind.

- Rebased.
- Squahed commits and reworded the commit message to follow the new guidelines.
- Made the `getdigits` change, in spite of my potestations lol.
- Fixed options.lua, the old commit didn't compile anymore. Changed `defaults={if_true={vi="ver:3,hor:6"}}` to `defaults={if_true="ver:3,hor:6"}`, I think `vi=` isn't a thing anymore? Also, added a `short_desc` (noticed the other options have one now).

zeertzjq commented on runtime/doc/options.txt:
```diff
@@ -4166,6 +4166,26 @@ A jump table for the options with a short description can be found at |Q_op|.
+	scrolling with a mouse. The option is a comma separated list of parts.
+	Each part consists of a direction and a count as follows:
+		direction:count,direction:count
+	Direction is one of either 'hor' or 'ver', 'hor' controls horizontal
+	scrolling and 'ver' controls vertical scrolling. Count sets the amount
```
Double quotes should be used here:
```suggestion
	Direction is one of either "hor" or "ver", "hor" controls horizontal
	scrolling and "ver" controls vertical scrolling. Count sets the amount
```

JaySandhu replied:
Made the change. Squashed the commit for cleaner repo history. Thanks for reviewing these old PRs btw 🙂.

zeertzjq commented:
CI is failing. Looks like this needs a rebase.

JaySandhu commented:
Rebased and addressed the main CI issue. The type of error messages was changed from `char_u*` to `char*` in https://github.com/neovim/neovim/commit/efa924f66b183d9cf2404ce91c4f009c27e0515a. This PR was still using `char_u*` so the compiler was complaining type mismatches.

I believe the remaining CI failures are unrelated (could be wrong though).

JaySandhu commented:
Rebased again. Fixed conflicts and linter complaints.

xiyaowong commented:
> you can no longer scroll inactive windows

Oh, no! Am I the only one who wants to scroll inactive windows?

zeertzjq commented:
> > you can no longer scroll inactive windows
> 
> Oh, no! Am I the only one who wants to scroll inactive windows?

Then just don't map scroll wheel keys.

xiyaowong commented:
@zeertzjq I'm sorry, I just tried the latest option and didn't understand it clearly, now I understand
